src/sdk/pynni/nni/compression/torch/pruning/amc/lib/net_measure.py

```diff
@@ -0,0 +1,123 @@
+# Copyright (c) Microsoft Corporation.
+# Licensed under the MIT license.
+
+import torch
+
+# [reference] https://github.com/ShichenLiu/CondenseNet/blob/master/utils.py
+
+
+def get_num_gen(gen):
+    return sum(1 for _ in gen)
+
+
+def is_leaf(model):
+    return get_num_gen(model.children()) == 0
+
+
+def get_layer_info(layer):
+    layer_str = str(layer)
+    type_name = layer_str[:layer_str.find('(')].strip()
+    return type_name
+
+
+def get_layer_param(model):
+    import operator
+    import functools
+
+    return sum([functools.reduce(operator.mul, i.size(), 1) for i in model.parameters()])
+
+count_ops = 0
+count_params = 0
+
+def measure_layer(layer, x):
+    global count_ops, count_params
+    delta_ops = 0
+    delta_params = 0
+    multi_add = 1
+    type_name = get_layer_info(layer)
+
+    # ops_conv
+    if type_name in ['Conv2d']:
+        out_h = int((x.size()[2] + 2 * layer.padding[0] - layer.kernel_size[0]) /
+                    layer.stride[0] + 1)
+        out_w = int((x.size()[3] + 2 * layer.padding[1] - layer.kernel_size[1]) /
+                    layer.stride[1] + 1)
+        delta_ops = layer.in_channels * layer.out_channels * layer.kernel_size[0] *  \
+                layer.kernel_size[1] * out_h * out_w / layer.groups * multi_add
+        delta_params = get_layer_param(layer)
+
+    # ops_nonlinearity
+    elif type_name in ['ReLU']:
+        delta_ops = x.numel() / x.size(0)
+        delta_params = get_layer_param(layer)
+
+    # ops_pooling
+    elif type_name in ['AvgPool2d']:
+        in_w = x.size()[2]
+        kernel_ops = layer.kernel_size * layer.kernel_size
+        out_w = int((in_w + 2 * layer.padding - layer.kernel_size) / layer.stride + 1)
+        out_h = int((in_w + 2 * layer.padding - layer.kernel_size) / layer.stride + 1)
+        delta_ops = x.size()[1] * out_w * out_h * kernel_ops
+        delta_params = get_layer_param(layer)
+
+    elif type_name in ['AdaptiveAvgPool2d']:
+        delta_ops = x.size()[1] * x.size()[2] * x.size()[3]
+        delta_params = get_layer_param(layer)
+
+    # ops_linear
+    elif type_name in ['Linear']:
+        weight_ops = layer.weight.numel() * multi_add
+        bias_ops = layer.bias.numel()
+        delta_ops = weight_ops + bias_ops
+        delta_params = get_layer_param(layer)
+
+    # ops_nothing
+    elif type_name in ['BatchNorm2d', 'Dropout2d', 'DropChannel', 'Dropout']:
+        delta_params = get_layer_param(layer)
+
+    # unknown layer type
+    else:
+        delta_params = get_layer_param(layer)
+
+    count_ops += delta_ops
+    count_params += delta_params
+
+    return
+
+
+def measure_model(model, H, W):
```

Comment:
is the `meauser_model` used during training/searching? I think this function is the same as `from nni.compression.torch.utils.counter import count_flops_params`

Reply:
Yes, meauser_model is used after training, a simpler version is used during training, they gave same results.
Just test 3 models, count_flops_params flops is higher than meauser_model for 3% - 8%

Reply:
`count_flops_params` is more precise?

Reply:
Yes, count_flops_params also count BN flops

Reply:
added TODO item to replace flops counter with count_flops_params in next PR.
The current flops counter is mixed with other collectors,  it trace (forward) once to collect both flops and feature maps.  Simply using count_flops_params will require additional trace.